types/wlr_workspace_v1.c

```diff
@@ -0,0 +1,544 @@
+#define _POSIX_C_SOURCE 200809L
+
+#include <assert.h>
+#include <string.h>
+#include <stdlib.h>
+#include <wlr/types/wlr_workspace_v1.h>
+#include <wlr/types/wlr_seat.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+
+#include "wlr-workspace-unstable-v1-protocol.h"
+
+#define WORKSPACE_V1_VERSION 1
+
+static void workspace_manager_idle_send_done(void *data) {
+	struct wlr_workspace_manager_v1 *manager = data;
+	struct wl_resource *resource, *tmp;
+	wl_resource_for_each_safe(resource, tmp, &manager->resources) {
+		zwlr_workspace_manager_v1_send_done(resource);
+	}
+
+	manager->idle_source = NULL;
+}
+
+static void workspace_manager_update_idle_source(
+		struct wlr_workspace_manager_v1 *manager) {
+	if (manager->idle_source) {
+		return;
+	}
+
+	manager->idle_source = wl_event_loop_add_idle(manager->event_loop,
+			workspace_manager_idle_send_done, manager);
+}
+
+static const struct zwlr_workspace_handle_v1_interface workspace_handle_impl;
+static struct wlr_workspace_handle_v1 *workspace_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+			&zwlr_workspace_handle_v1_interface,
+			&workspace_handle_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void workspace_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void workspace_handle_activate(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_workspace_handle_v1 *workspace = workspace_from_resource(resource);
+	if (!workspace) {
+		return;
+	}
+
+	workspace->pending |= WLR_WORKSPACE_HANDLE_V1_STATE_ACTIVE;
+}
+
+static void workspace_handle_deactivate(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_workspace_handle_v1 *workspace = workspace_from_resource(resource);
+	if (!workspace) {
+		return;
+	}
+
+	workspace->pending &= ~WLR_WORKSPACE_HANDLE_V1_STATE_ACTIVE;
+}
+
+static void workspace_handle_resource_destroy(struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static bool fill_array_from_workspace_state(struct wl_array *array,
```

Comment:
What does this function do?

Reply:
Exactly what it says. It takes a state which is a bitmask of the workspace state, and fills an array which contains the set bits (array is needed because that's what the protocol has).

Reply:
I see. thnaks